### PR TITLE
Add automated building & release of kerf1 for macOS (Linux broken)

### PR DESCRIPTION
**Pull request by luxemboye**
This PR adds complete support for automated building and (on tags) release of kerf1 for macOS. (and partial Linux)

## macOS works - Linux errors out
kerf1 seems to be built with macOS in mind: it uses clang, and builds flawlessly under macOS (clang+GNU ld).

To build under Linux, a couple of changes have to be done:
- an include to `sys/macros.h` is missing for `wire.c` for Linux
- `libedit` seems to be macOS specific (see [`Makefile` comment l.39](https://github.com/kevinlawler/kerf1/blob/master/src/Makefile#L39)). As such, the [macports version](https://ports.macports.org/port/libedit/) was taken and built manually inside of the workflow.

This doesn't seem to suffice though, as it still errors out.
If both of these steps are taken inside of a [GitHub-provided Codespaces VM](https://github.com/features/codespaces), the build succeeds, but the final binary errors out at run-time, complaining about the libedit library file.

I have tested around and modified the source code with a friend, and these things might hinder the build:
1. possibly the library not being found (giving the exact path `/usr/local/lib/libedit.a` seems to help a bit in the Makefile instead of `-ledit`) This seems to be somewhat confirmed through [the comment (l.39) about "self use"](https://github.com/kevinlawler/kerf1/blob/master/src/Makefile#L39), but no .a file is to be found. @kevinlawler plz fix the broken linux build, it's not clear where that .a file comes from or how it should be built from source
2. including kerf.h instead of lz4.h inside of lz4.c tself. If that is fixed, it needs adding some includes and typedefs at the [`extern "C"`, `Vesion` part](https://github.com/kevinlawler/kerf1/blob/master/src/lz4.h#L49) (inside if lz4.h)
3. a [compiler option inside of lz4.c](https://github.com/kevinlawler/kerf1/blob/master/src/lz4.c#L82) that's commented, about stdc version

The usage of termcap doesn't help either: on Arch Linux for example, it finds itself [in the AUR](https://aur.archlinux.org/packages/termcap) because the maintainers don't seem to consider it useful/up to todays standards anymore.

Linux is being left in, as only small changes have to be made for it to work. For example, if the Makefile is modified on the tree and the library added there, only a removal of the manual building of the library is needed.

----
## Workflow - Automatic building and releases
To sum up the workflow:
- all push to master:
   - builds and tests kerf1
   - uploads download binaries/artifacts (30 day expiration max and by default i think)
- all tags:
   - builds and tests kerf1
   - uploads download binaries/artifacts (30 day expiration max and by default i think)
   - creates a **draft release**, EVEN IF the builds fail

As such, even if the linux build is broken, a draft release is created.
This also means that releases have to be tagged, and that the owner of the repo still has to fill in the body of the release and publish it, but the executables are already present. This is to avoid un-wanted tags to be published.

## Tldr
- macOS works, building seems to be broken for Linux, probably specific libedit library needed by @kevinlawler 
- Draft releases are created, the owner just needs to publish them (to avoid mis-publish)

See the (published) release by yourself: https://github.com/luxemboye/kerf1/releases/tag/v0.0.1
Made by this run: https://github.com/luxemboye/kerf1/actions/runs/2724745156

### CODE REVIEW

**Comment by luxemboye** [comment]
Yes the workflow run technically shows "errors" (that are ignored inside of the action **on purpose**), but someof these errors will completely disappear once Linux is fixed, and possibly Windows is added.

Sadly, the testing step error cannot be removed unless the testing binary gets a flag that makes it exit on execution

**Comment by kevinlawler** [comment]
When Kerf was closed-sourced it used a static compilation with `libedit.a` to get around inconsistencies between different Linux releases with respect to `-ledit` in order to release a single Linux-universal binary. Now that Kerf is open-source my assumption is this probably doesn't matter any more. I could add the `libedit.a` to this repo but it would probably be better to have any `libedit.a` code removed and replaced with a `-ledit` (or whatever is right these days) compiler flag for Linux. Looking at the Makefile, it seems like correct line is commented and the correct line is uncommented for Linux builds, so it isn't immediately apparent to me what the issue is there.

**Comment by kevinlawler** [comment]
I don't see your contact info. Could you contact me at kerf.concerns@gmail.com?

**Comment by luxemboye** [comment]
> (...) it would probably be better to have any `libedit.a` code removed and replaced with a `-ledit` (or whatever is right these days) compiler flag for Linux. (...)

It would! and -ledit seems to be right.

I'd guess it's a **linker issue**: it compiles fine under macOS, Linux and for that matter even FreeBSD (with some tiny changes), but it **doesn't link "properly"** under **Linux** (errors) nor FreeBSD (silently failing with no output and no binaries). Both of these have GNU linkers.
I tried building on Alpine (clang/musl), but faced different compiler issues that I didn't consider worth fixing comparatively.

There are 3 big editlines:
- Minix editline (["call compatible with the FSF readline library"](https://troglobit.com/projects/editline/))
- GNU readline
- NetBSD editline ([macOS 10.14 seems to have some custom half-broken version](https://github.com/knz/go-libedit#macososx-due-to-restrictions-due-to-changes-macos-mojave))

Even when using the **macOS editline** (NetBSD one) under **Linux**, the GNU linker **errors out**. I even tried using mold and lld for fun (both compatible with GNU ld), but both failed (see examples).
As such, I think it might be a **difference** in how the **macOS linker** (for Mach-O) deals with duplicate symbols compared to the **GNU-compatible linkers** (for ELFs). It's probably some kind of "self-referencing includes", possibly something to do with `kerf.h`, but that's just a simple guess. I did play around with the idea of getting rid of those, but didn't get anywhere far enough.

Some linker versions with examples:
```
GNU ld (GNU Binutils) 2.38

/usr/bin/ld: obj/cow.o:(.bss+0x0): multiple definition of `size_of_type_element'; obj/alter.o:(.bss+0xa20): first defined here
```
```
LLD 14.0.6 (compatible with GNU linkers)

ld.lld: error: duplicate symbol: size_of_type_element
>>> defined at alter.c
>>>            obj/alter.o:(size_of_type_element)
>>> defined at cow.c
>>>            obj/cow.o:(.bss+0x0)
```
```
mold 1.3.1 (7f7d10cc896e0c7513d09723d5ca3d8346bdea49; compatible with GNU ld)

mold: error: duplicate symbol: obj/cow.o: obj/alter.o: size_of_type_element
```

---
> I don't see your contact info. Could you contact me at [kerf.concerns@gmail.com](mailto:kerf.concerns@gmail.com)?

I just sent you an email!

**Comment by jibanes** [comment]
> mold: error: duplicate symbol: obj/cow.o: obj/alter.o: size_of_type_element

use -fcommon in your CFLAGS

**Comment by luxemboye** [comment]
> > mold: error: duplicate symbol: obj/cow.o: obj/alter.o: size_of_type_element
> 
> use -fcommon in your CFLAGS

That's a pretty good idea.
The linker doesn't complain anymore, but when testing/executing, it errors out, not able to find the shared library, and gives an exit code `127`.

See this workflow run on a modified `Makefile` with `-fcommon` in the `CFLAGS`. The "Test kerf" step which technically "passes" even though it errors out, thats because I haven't yet implemented expect to automate the exiting of kerf. :| I'll fix that once I have a working binary.
https://github.com/luxemboye/kerf1/runs/7486852077

The specific error:
```
$ ./kerf_test 
./kerf_test: error while loading shared libraries: libedit.so.0: cannot open shared object file: No such file or directory
```

It links against a library but then can't find it anymore? o_0
When checking ldconfig, it seems like only libedit.so.2 is known...
```
$ ldconfig -p | grep edit
        libedit.so.2 (libc6,x86-64) => /lib/x86_64-linux-gnu/libedit.so.2
```

A very dirty hack (don't do this), and it seems to work:
```
$ cd /lib/x86_64-linux-gnu/
$ sudo ln -s libedit.so.2 libedit.so.0
$ cd /workspaces/kerf1/src/
$ ./kerf_test 
 ╔═════════════════════════════════════════════════════════════════════════╗
 ║                                                                         ║
 ║                                                                         ║
 ║                                                                         ║
 ║                     ██╗  ██╗ █████╗ ██████╗  ██████╗                    ║
 ║                     ██║ ██╔╝██╔══██╗██╔══██╗███╔═══╝                    ║
 ║                     █████╔╝ ██████╔╝██████╔╝██████╗                     ║
 ║                     ██╔═██╗ ██╔═══╝ ██╔══██╗██╔═══╝                     ║
 ║                     ██║  ██╗╚██████╗██║  ██║██║                         ║
 ║                     ╚═╝  ╚═╝ ╚═════╝╚═╝  ╚═╝╚═╝     ©                   ║
 ║                                                                         ║
 ║                                                                         ║
 ║                                  {SOLO, 64-BIT, DEBUG, -O0, 2022.07.24} ║
 ╚═════════════════════════════════════════════════════════════════════════╝

Testing case   0...
Testing case 100...
Testing case 200...
Testing case 300...
Testing case 400...
Testing case 500...
Test pass rate: 1.0000, Total: 554, Pass: 553, Skip: 1, Fail: 0, Time: 0.497262s
Passed
KeRF> quit

[DEBUG] Testing for leaked memory...
[DEBUG] Done...
[DEBUG] OK: Done OK.
```

So now, I'm guessing the issue lies with which library it links to, and which one the linker uses at run-time.
This is confirmed after deleting that dirty hack, and using `sudo ldconfig -v` to re-create the links.
Just an update, I'm looking into a cleaner solution.

**Comment by kevinlawler** [comment]
> The specific error:
> 
> ```
> $ ./kerf_test 
> ./kerf_test: error while loading shared libraries: libedit.so.0: cannot open shared object file: No such file or directory
> ```
> 
> It links against a library but then can't find it anymore? o_0 When checking ldconfig, it seems like only libedit.so.2 is known...
> 
> ```
> $ ldconfig -p | grep edit
>         libedit.so.2 (libc6,x86-64) => /lib/x86_64-linux-gnu/libedit.so.2
> ```
> 
> A very dirty hack (don't do this), and it seems to work:
> 
> ```
> $ cd /lib/x86_64-linux-gnu/
> $ sudo ln -s libedit.so.2 libedit.so.0
> 
> So now, I'm guessing the issue lies with which library it links to, and which one the linker uses at run-time. This is confirmed after deleting that dirty hack, and using `sudo ldconfig -v` to re-create the links. Just an update, I'm looking into a cleaner solution.

This looks remarkably similar to the original issues we had building cross-Linux binaries where the executable would look for libedit.so in two different places depending on what flavor of Linux it was linked for. Why this would be happening on your build I'm not exactly sure. Is it possible that you are linking for one variety of Linux and then deploying the binary to another?

**Comment by luxemboye** [comment]
> Why this would be happening on your build I'm not exactly sure. Is it possible that you are linking for one variety of Linux and then deploying the binary to another?
It linked to the manually (as in, not distro provided) installed `libedit.so.0`, which is not in the ldconfig's (linker runtime) path.

Tl;dr, linker (after compile) finds a dynamic library, but it's not in the runtime's path. From what I know, the runtime one doesn't look at all paths, but only the registered libraries, which can be regenerated with `ldconfig`.

I got it to work properly over on my `staging` branch, but right now, before adding that to my `master` branch, I'm working on an `expect` script to properly test it instead of just timing out as it is right now.

**Comment by luxemboye** [comment]
Boom, this should be it. There you go. Binaries built on Ubuntu (20.04) and for macOS (11.6).

Check the action on my master branch (incl. artifacts): https://github.com/luxemboye/kerf1/actions/runs/2727257182

The annotations are due to errors when it doesn't find the windows artifact, though I could remove it seeing the state of #11.

**Comment by luxemboye** [comment]
To sum up what this PR does:

For **_all_ pushes** on the `master` branch:
1. **build** binaries for Ubuntu and macOS, using system libraries
2. **test** (run debug binary) using expect
3. **upload binaries** as artifacts on the workflow run (downloadable for a limited amount of time until they expire)

For tags, it:
4. creates **draft release** (**binaries** are **already attached**. The maintainer simply has to fill in text.)

**Comment by tonetheman** [comment]
That expect is pretty cool. I was looking at it on my own wondering how to fix that. Good stuff.

**Comment by kevinlawler** [comment]
yes, it is cool

probably better even than doing the compiler flag thing he mentioned
earlier and hard-coding it to exit
